sw/airborne/modules/calibration/mag_calib_ukf.c

```diff
@@ -0,0 +1,200 @@
+
+#define PRINT(string,...) fprintf(stderr, "[CALIB_UKF->%s()] " string,__FUNCTION__ , ##__VA_ARGS__)
+
+#if !defined MAG_CALIB_UKF_GEO_MAG_TIMEOUT
+#define MAG_CALIB_UKF_GEO_MAG_TIMEOUT 0
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_GEO_MAG_TIMEOUT)
+
+#if !defined MAG_CALIB_UKF_NORM
+#define MAG_CALIB_UKF_NORM 1.0f
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_NORM)
+
+#if !defined MAG_CALIB_UKF_NOISE_RMS
+#define MAG_CALIB_UKF_NOISE_RMS 2e-1f
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_NOISE_RMS)
+
+#if !defined MAG_CALIB_UKF_HOTSTART
+#define MAG_CALIB_UKF_HOTSTART TRUE
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_HOTSTART)
+
+#if !defined MAG_CALIB_UKF_HOTSTART_SAVE_FILE
+#define MAG_CALIB_UKF_HOTSTART_SAVE_FILE /data/ftp/internal_000/mag_ukf_calib.txt
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_HOTSTART_SAVE_FILE)
+
+#if !defined MAG_CALIB_UKF_VERBOSE || !MAG_CALIB_UKF_VERBOSE
+#define VERBOSE_PRINT(...)
+#else
+#define VERBOSE_PRINT PRINT
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_VERBOSE)
+
+TRICAL_instance_t mag_calib;
+static abi_event mag_ev;
+
+time_t mag_field_from_geo_mag;
+bool   update_geo_mag_field = true;
+struct FloatVect3 H = { .x = MAG_CALIB_UKF_NORM, .y = 0.0f, .z =  0.0f};
+
+#if MAG_CALIB_UKF_HOTSTART
+static FILE *fp;
+char hotstart_file_name[512];
+#endif
+
+void mag_calib_ukf_init(void)
+{
+#if USE_MAGNETOMETER
+  TRICAL_init(&mag_calib);
+  TRICAL_norm_set(&mag_calib, MAG_CALIB_UKF_NORM);
+  TRICAL_noise_set(&mag_calib, MAG_CALIB_UKF_NOISE_RMS);
+
+  H.x = AHRS_H_X;
+  H.y = AHRS_H_Y;
+  H.z = AHRS_H_Z;
+  VERBOSE_PRINT("Local magnetic field loaded from airframe file (Hx: %4.2f, Hy: %4.2f, Hz: %4.2f)\n", H.x, H.y, H.z);
+
+  mag_calib_hotstart_read();
+  AbiBindMsgIMU_MAG_INT32(IMU_BOARD_ID, &mag_ev, mag_calib_ukf_run);
+#endif
+}
+
+void mag_calib_ukf_run(uint8_t __attribute__((unused)) sender_id, uint32_t __attribute__((unused)) stamp, struct Int32Vect3 *mag)
+{
+  float measurement[3] = {0.0, 0.0, 0.0}, calibrated_measurement[3] = {0.0, 0.0, 0.0};
+  /** Update geo_mag based on MAG_CALIB_UKF_GEO_MAG_TIMEOUT (0 = no periodic updates) **/
+  if (update_geo_mag_field && geo_mag.ready) {
+    double n                = double_vect3_norm(&geo_mag.vect);
+    if (n > 0.01) {
+      H.x                     = (float) geo_mag.vect.x / n;
+      H.y                     = (float) geo_mag.vect.y / n;
+      H.z                     = (float) geo_mag.vect.z / n;
+      mag_field_from_geo_mag  = time(0);
+      update_geo_mag_field    = false;
+      VERBOSE_PRINT("Updating local magnetic field from geo_mag module (Hx: %4.2f, Hy: %4.2f, Hz: %4.2f)\n", H.x, H.y, H.z);
+    }
+  }
+  if (MAG_CALIB_UKF_GEO_MAG_TIMEOUT && GpsFixValid() && difftime(time(0), mag_field_from_geo_mag) >= MAG_CALIB_UKF_GEO_MAG_TIMEOUT) {
+    geo_mag.ready           = false;
+    geo_mag.calc_once       = true;   ///< Geo_mag will not re-update the calculation when the throttle is on so this is neccesary
```

Comment:
@gautierhattenberger Should the `geo_mag` timed updates be transferred to the `geo_mag` module instead of being triggered here? If not should I make the updates a periodical function in this module instead of checking for the time here?